<a href="https://colab.research.google.com/github/pankajr141/experiments/blob/master/Reasoning/timeseries/grangercausality%20-%20causal%20impact%20analysis%20on%20timeseries.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In this notebook we will try to see if any feature has any causal impact on other feature. In short can we use 1 feature to predict future values of other feature

---

Download the dataset from https://www.kaggle.com/datasets/sudalairajkumar/cryptocurrencypricehistory

In [ ]:
!unzip archive.zip

Archive:  archive.zip
  inflating: coin_Aave.csv           
  inflating: coin_BinanceCoin.csv    
  inflating: coin_Bitcoin.csv        
  inflating: coin_Cardano.csv        
  inflating: coin_ChainLink.csv      
  inflating: coin_Cosmos.csv         
  inflating: coin_CryptocomCoin.csv  
  inflating: coin_Dogecoin.csv       
  inflating: coin_EOS.csv            
  inflating: coin_Ethereum.csv       
  inflating: coin_Iota.csv           
  inflating: coin_Litecoin.csv       
  inflating: coin_Monero.csv         
  inflating: coin_NEM.csv            
  inflating: coin_Polkadot.csv       
  inflating: coin_Solana.csv         
  inflating: coin_Stellar.csv        
  inflating: coin_Tether.csv         
  inflating: coin_Tron.csv           
  inflating: coin_USDCoin.csv        
  inflating: coin_Uniswap.csv        
  inflating: coin_WrappedBitcoin.csv  
  inflating: coin_XRP.csv            


In [ ]:
import os
import pandas as pd

dfA = None
for filename in os.listdir('.'):
  if not filename.endswith('.csv'):
    continue
  df = pd.read_csv(filename)
  closing_colname = 'close_{}'.format(df['Symbol'].tolist()[0])
  df[closing_colname] = df['Close']
  df = df[['Date', closing_colname]]
  if dfA is None:
    dfA = df
    continue

  dfA = pd.merge(dfA, df, on=['Date'], how='inner')

dfA

Date  close_DOGE  close_TRX   close_XMR    close_ETH  \
0    2020-10-05 23:59:59    0.002612   0.026443  112.903248   353.956772   
1    2020-10-06 23:59:59    0.002611   0.025580  106.665996   340.815845   
2    2020-10-07 23:59:59    0.002595   0.025477  109.989031   341.808669   
3    2020-10-08 23:59:59    0.002596   0.025671  110.909933   350.766159   
4    2020-10-09 23:59:59    0.002653   0.026223  116.034673   365.590477   
..                   ...         ...        ...         ...          ...   
270  2021-07-02 23:59:59    0.245264   0.066512  211.345143  2150.040364   
271  2021-07-03 23:59:59    0.246411   0.066813  214.752778  2226.114282   
272  2021-07-04 23:59:59    0.246483   0.067298  219.841418  2321.724112   
273  2021-07-05 23:59:59    0.231614   0.064574  214.464993  2198.582464   
274  2021-07-06 23:59:59    0.234422   0.064935  222.137136  2324.679449   

     close_XLM     close_BTC  close_LINK  close_USDC  close_SOL  ...  \
0     0.073419  10793.339428    9.623703    1.000472   2.657644  ...   
1     0.072501  10604.405889    8.779272    1.000883   2.176250  ...   
2     0.072301  10668.968955    8.957850    1.000168   2.291796  ...   
3     0.073095  10915.685731    9.508818    1.000431   2.365220  ...   
4     0.075144  11064.457592   10.346902    1.000727   2.539859  ...   
..         ...           ...         ...         ...        ...  ...   
270   0.263427  33897.048590   18.283561    1.000035  34.020482  ...   
271   0.264324  34668.548402   18.538818    0.999984  34.478816  ...   
272   0.268128  35287.779766   19.262838    0.999500  34.310601  ...   
273   0.254456  33746.002456   18.371589    1.000528  32.984588  ...   
274   0.260190  34235.193451   20.080466    1.000059  34.269140  ...   

     close_UNI  close_USDT  close_EOS  close_AAVE  close_DOT  close_MIOTA  \
0     3.419205    1.000866   2.529068   53.219243   4.172700     0.271726   
1     2.834241    1.000448   2.693054   42.401599   3.756162     0.253433   
2     2.800064    1.000707   2.646043   40.083976   3.858798     0.256832   
3     3.270306    1.001252   2.609671   43.764463   4.102642     0.262097   
4     3.097367    1.001180   2.645295   46.817744   4.262468     0.271023   
..         ...         ...        ...         ...        ...          ...   
270  18.241726    1.000033   3.934325  233.948437  15.340774     0.828754   
271  19.524757    0.999956   4.046421  259.655225  15.553716     0.829169   
272  20.771874    0.999501   4.059930  277.038792  16.014026     0.844145   
273  20.095181    1.000667   3.819266  307.829079  15.235582     0.820022   
274  22.400186    1.000090   3.882325  316.898507  16.143564     0.851735   

     close_ATOM  close_XEM  close_CRO  close_XRP  
0      5.382198   0.122335   0.150009   0.250816  
1      4.732915   0.117298   0.147198   0.245341  
2      4.886457   0.117939   0.147583   0.248539  
3      5.045788   0.119316   0.150626   0.251454  
4      5.157042   0.119971   0.152428   0.253491  
..          ...        ...        ...        ...  
270   11.764925   0.131455   0.113914   0.656763  
271   11.817927   0.134756   0.118407   0.672888  
272   12.263548   0.137617   0.122427   0.694945  
273   13.548475   0.130602   0.115814   0.654300  
274   13.159371   0.133591   0.122856   0.665402  

[275 rows x 24 columns]

### Granger Causality

https://en.wikipedia.org/wiki/Granger_causality

Ability of a timeseries to forecast another timeseries. Its a Statistical causality test which provide a way to investigate causality in term of prediction. It look at power of prediction.

can I enhance the enhance the estimation of y by looking at the past values of x and y ?


---


Lets take 2 coins BTC (bitcoin) and LTC (litecoin) and see if litecoin has any statistical impact on prediction on future values of BTC 

We will use **Granger Causality** present in statsmodels package for this purpose.

However we cannot use granger causality directly with out data, there are condition to check before generating granger causality
1. series should be stationary

In [ ]:
from statsmodels.tsa.stattools import adfuller


def _is_series_stationary(df, col, differencing_order):
  result = adfuller(df[col])
  p_value = result[1]
  stationary = False
  if p_value <= 0.05:
    stationary = True

  print("column: {}, do: {}, Test Statistics:{}, p-value: {}, stationary: {}".format(col, differencing_order, result[0], p_value, stationary))
  return stationary

# columns = ['close_BTC', 'close_LTC']
for col in dfA.columns:
  if col in ['Date']:
    continue
  differencing_order = 0
  while not _is_series_stationary(dfA, col, differencing_order):
    differencing_order += 1

    # Exit if even 2nd order differential is not stationary
    if differencing_order > 2:
      print("Unable to make column: {} stationary".format(col))
      break
    dfA[col] = dfA[col].diff().fillna(0)

column: close_DOGE, do: 0, Test Statistics:-1.1431758403487646, p-value: 0.6975505191745451, stationary: False
column: close_DOGE, do: 1, Test Statistics:-4.58986040661769, p-value: 0.00013484380258368858, stationary: True
column: close_TRX, do: 0, Test Statistics:-1.4952966636057972, p-value: 0.5358367158181646, stationary: False
column: close_TRX, do: 1, Test Statistics:-4.051794195716015, p-value: 0.0011623102555727808, stationary: True
column: close_XMR, do: 0, Test Statistics:-1.6798784024482698, p-value: 0.4415084057650066, stationary: False
column: close_XMR, do: 1, Test Statistics:-22.249170973594357, p-value: 0.0, stationary: True
column: close_ETH, do: 0, Test Statistics:-1.3250662464899745, p-value: 0.617627892675766, stationary: False
column: close_ETH, do: 1, Test Statistics:-5.336917794257202, p-value: 4.589492864385697e-06, stationary: True
column: close_XLM, do: 0, Test Statistics:-1.6908520667756592, p-value: 0.4358318237750384, stationary: False
column: close_XLM, do:

In [ ]:
from statsmodels.tsa.stattools import grangercausalitytests
result = grangercausalitytests(dfA[['close_BTC', 'close_LTC']], maxlag=3)  # upto lag 3


Granger Causality
number of lags (no zero) 1
ssr based F test:         F=5.9169  , p=0.0156  , df_denom=271, df_num=1
ssr based chi2 test:   chi2=5.9824  , p=0.0144  , df=1
likelihood ratio test: chi2=5.9181  , p=0.0150  , df=1
parameter F test:         F=5.9169  , p=0.0156  , df_denom=271, df_num=1

Granger Causality
number of lags (no zero) 2
ssr based F test:         F=3.2389  , p=0.0407  , df_denom=268, df_num=2
ssr based chi2 test:   chi2=6.5987  , p=0.0369  , df=2
likelihood ratio test: chi2=6.5202  , p=0.0384  , df=2
parameter F test:         F=3.2389  , p=0.0407  , df_denom=268, df_num=2

Granger Causality
number of lags (no zero) 3
ssr based F test:         F=2.8286  , p=0.0390  , df_denom=265, df_num=3
ssr based chi2 test:   chi2=8.7099  , p=0.0334  , df=3
likelihood ratio test: chi2=8.5733  , p=0.0355  , df=3
parameter F test:         F=2.8286  , p=0.0390  , df_denom=265, df_num=3


since p value of test is p=0.0156 which is much lower then 0.05 that means the result is statistically significant or LTC has impact on future values of BTC.

In [ ]:
result

{1: ({'ssr_ftest': (11.017711715643845, 0.0010261403883030577, 271.0, 1),
   'ssr_chi2test': (11.139679004008906, 0.0008450051269438997, 1),
   'lrtest': (10.919189231218297, 0.0009517325248948702, 1),
   'params_ftest': (11.017711715643859, 0.0010261403883030248, 271.0, 1.0)},
   array([[0., 1., 0.]])]),
 2: ({'ssr_ftest': (7.931369526422364, 0.000450337659811857, 268.0, 2),
   'ssr_chi2test': (16.158685676964964, 0.00030987460618830503, 2),
   'lrtest': (15.69854494098945, 0.0003900356274821154, 2),
   'params_ftest': (7.931369526422312, 0.0004503376598118783, 268.0, 2.0)},
   array([[0., 0., 1., 0., 0.],
          [0., 0., 0., 1., 0.]])]),
 3: ({'ssr_ftest': (5.285765770414507, 0.00148383639981008, 265.0, 3),
   'ssr_chi2test': (16.27616931569146, 0.000995317700852813, 3),
   'lrtest': (15.807790164213657, 0.0012416545919022005, 3),
   'params_ftest': (5.28576577041447, 0.0014838363998101532, 265.0, 3.0)},
   array([[0., 0., 0., 1., 0., 0., 0.],
          [0., 0., 0., 0., 1., 0., 0.

### What other feature can help predict future values of BTC ? 



In [ ]:
columns_stats = []
for column in dfA.columns.tolist():
  if column in  ['Date', 'close_BTC']:
    continue

  max_lag = 3    # search upto max_lags
  result = grangercausalitytests(dfA[['close_BTC', column]], maxlag=max_lag)
  print(result)

  min_f_score = 1
  lag_num_with_min_f_score = -1
  column_included_in_forecast = False

  for i in range(1, max_lag + 1):
    f_score = result[i][0]['ssr_ftest'][1]
    if f_score < min_f_score and f_score < 0.05:
      column_included_in_forecast = True

    if f_score < min_f_score:
      min_f_score = f_score
      lag_num_with_min_f_score = i

  columns_stats.append({'col_name': column, 'column_included_in_forecast': column_included_in_forecast, 
                        'f_score': min_f_score, 'lag_num_with_min_f_score': lag_num_with_min_f_score})


Granger Causality
number of lags (no zero) 1
ssr based F test:         F=7.2467  , p=0.0075  , df_denom=271, df_num=1
ssr based chi2 test:   chi2=7.3269  , p=0.0068  , df=1
likelihood ratio test: chi2=7.2307  , p=0.0072  , df=1
parameter F test:         F=7.2467  , p=0.0075  , df_denom=271, df_num=1

Granger Causality
number of lags (no zero) 2
ssr based F test:         F=4.5278  , p=0.0116  , df_denom=268, df_num=2
ssr based chi2 test:   chi2=9.2246  , p=0.0099  , df=2
likelihood ratio test: chi2=9.0722  , p=0.0107  , df=2
parameter F test:         F=4.5278  , p=0.0116  , df_denom=268, df_num=2

Granger Causality
number of lags (no zero) 3
ssr based F test:         F=3.0451  , p=0.0293  , df_denom=265, df_num=3
ssr based chi2 test:   chi2=9.3765  , p=0.0247  , df=3
likelihood ratio test: chi2=9.2185  , p=0.0265  , df=3
parameter F test:         F=3.0451  , p=0.0293  , df_denom=265, df_num=3
{1: ({'ssr_ftest': (7.246707843014532, 0.00754522570705606, 271.0, 1), 'ssr_chi2test': (7.3269

In [ ]:
pd.set_option('display.float_format', lambda x: '%.3f' % x)
pd.DataFrame(columns_stats)

col_name  column_included_in_forecast  f_score  \
0     close_XLM                         True    0.008   
1    close_LINK                         True    0.000   
2    close_AAVE                         True    0.049   
3     close_XMR                         True    0.000   
4    close_USDT                        False    0.487   
5     close_SOL                         True    0.001   
6     close_LTC                         True    0.000   
7     close_ADA                         True    0.021   
8    close_DOGE                         True    0.008   
9    close_ATOM                         True    0.010   
10    close_XRP                         True    0.002   
11    close_DOT                         True    0.004   
12    close_ETH                         True    0.000   
13   close_USDC                        False    0.078   
14    close_EOS                         True    0.000   
15    close_BNB                         True    0.002   
16    close_UNI                        False    0.165   
17    close_CRO                         True    0.001   
18   close_WBTC                        False    0.120   
19  close_MIOTA                         True    0.003   
20    close_TRX                         True    0.002   
21    close_XEM                         True    0.008   

    lag_num_with_min_f_score  
0                          1  
1                          2  
2                          2  
3                          2  
4                          1  
5                          2  
6                          2  
7                          2  
8                          1  
9                          3  
10                         2  
11                         2  
12                         2  
13                         1  
14                         2  
15                         3  
16                         1  
17                         2  
18                         1  
19                         2  
20                         2  
21                         2

### Transfer Entropy

transfer entropy is a non-parametric statistic measuring the amount of directed transfer of information between 2 random processes. It quantifies the statistical coherence between system evolving in time.

the entropy of a random variable is the average level of "information", "surprise", or "uncertainty" inherent to the variable's possible outcomes.
Since every feature has some entropy, our aim is to minimise it.
https://en.wikipedia.org/wiki/Entropy_(information_theory)

Can I reduce the entropy/uncertainty of in future values of y by looking at the past values of x and y ?

### Difference transfer entropy vs granger causality

**transfer entropy is**

1. Autoregressive approach to causal inference
1. more predice
2. more visually interpretable
3. Model free

**granger causality is**

1. information theoretic approach to causal inference 
1. Easier to automate
2. Less computationaly expensive
3. Easier to interpret the meaning of values obtained

**both approaches can** 
1. fail to distinguish causal interaction from correlation that arise due to similar governing equation or correlation that are induced by presence of common external forcing 
2. when system consist of more then 2 interacting processes they can return fake causality, i.e fail to distingish between direct and indirect causal interaction.


### References

1. https://www.youtube.com/watch?v=focIC0v5Rds
